File: examples/grpc-client/javascript/tasks.js
#!/usr/bin/env node

/**
 * KeystoneDB gRPC JavaScript Client Example: Task Management
 *
 * Demonstrates remote database operations using the JavaScript gRPC client.
 */

const path = require('path');

// Import the KeystoneDB client
const clientPath = path.join(__dirname, '../../../bindings/javascript/client/dist');
let KeystoneClient;

try {
    const module = require(clientPath);
    KeystoneClient = module.KeystoneClient || module.default?.KeystoneClient;

    if (!KeystoneClient) {
        console.error('Error: KeystoneClient not found in module');
        console.error('Available exports:', Object.keys(module));
        process.exit(1);
    }
} catch (error) {
    console.error('Error importing KeystoneDB client:', error.message);
    console.error('\nMake sure you\'ve built the JavaScript client:');
    console.error('  cd bindings/javascript/client');
    console.error('  npm install');
    console.error('  npm run build');
    process.exit(1);
}

const SERVER_ADDR = 'localhost:50051';

async function createTasks(client) {
    console.log('--- Creating Tasks ---');

    const tasks = [
        {
            id: 'task#1',
            project: 'project#backend',
            title: 'Implement user authentication',
            description: 'Add JWT-based auth system',
            status: 'in-progress',
            priority: 'high',
        },
        {
            id: 'task#2',
            project: 'project#backend',
            title: 'Set up database migrations',
            description: 'Create migration scripts',
            status: 'pending',
            priority: 'medium',
        },
        {
            id: 'task#3',
            project: 'project#frontend',
            title: 'Design login page',
            description: 'Create UI mockups',
            status: 'completed',
            priority: 'low',
        },
    ];

    for (const task of tasks) {
        // Build item attributes
        const attributes = {
            title: { S: task.title },
            description: { S: task.description },
            status: { S: task.status },
            priority: { S: task.priority },
            created: { N: String(Math.floor(Date.now() / 1000)) },
        };

        // Put task by ID
        await client.put({
            partitionKey: Buffer.from(task.id),
            item: { attributes },
        });

        // Also put with project partition for querying
        await client.put({
            partitionKey: Buffer.from(task.project),
            sortKey: Buffer.from(task.id),
            item: { attributes },
        });

        console.log(`✅ Created task: ${task.id}`);
    }
}

async function getTask(client) {
    console.log('\n--- Retrieving Task ---');

    const response = await client.get({
        partitionKey: Buffer.from('task#1'),
    });

    if (response.item) {
        console.log('Task task#1:');
        printItem(response.item);
    } else {
        console.log('Task not found');
    }
}

async function queryTasks(client) {
    console.log('\n--- Querying Tasks by Project ---');

    const response = await client.query({
        partitionKey: Buffer.from('project#backend'),
        limit: 10,
    });

    console.log(`Found ${response.items.length} tasks for project#backend`);

    response.items.forEach((item, i) => {
        console.log(`\nTask ${i + 1}:`);
        printItem(item);
    });
}

async function batchOperations(client) {
    console.log('\n--- Batch Operations ---');

    const response = await client.batchGet({
        keys: [
            { partitionKey: Buffer.from('task#1') },
            { partitionKey: Buffer.from('task#2') },
        ],
    });

    console.log(`Retrieved ${response.items.length} tasks in batch operation`);
}

async function deleteTask(client) {
    console.log('\n--- Deleting Task ---');

    // Delete task#3
    await client.delete({
        partitionKey: Buffer.from('task#3'),
    });

    // Also delete from project partition
    await client.delete({
        partitionKey: Buffer.from('project#frontend'),
        sortKey: Buffer.from('task#3'),
    });

    console.log('✅ Deleted task#3');
}

function printItem(item) {
    if (!item.attributes) return;

    for (const [key, value] of Object.entries(item.attributes)) {
        let valStr;
        if (value.S !== undefined) {
            valStr = value.S;
        } else if (value.N !== undefined) {
            valStr = value.N;
        } else if (value.Bool !== undefined) {
            valStr = String(value.Bool);
        } else {
            valStr = JSON.stringify(value);
        }
        console.log(`  ${key}: ${valStr}`);
    }
}

async function main() {
    console.log(`Connecting to KeystoneDB server at ${SERVER_ADDR}...`);

    const client = new KeystoneClient(SERVER_ADDR);

    try {
        console.log('✅ Connected successfully!\n');

        await createTasks(client);
        await getTask(client);
        await queryTasks(client);
        await batchOperations(client);
        await deleteTask(client);

        console.log('\n✅ All operations completed successfully!');
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    } finally {
        client.close();
    }
}

// Run if executed directly
if (require.main === module) {
    main().catch((error) => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}

File: examples/grpc-client/python/tasks.py
#!/usr/bin/env python3
"""
KeystoneDB gRPC Python Client Example: Task Management

Demonstrates remote database operations using the Python gRPC client.
"""

import sys
import time
from pathlib import Path

# Add bindings to path
sys.path.insert(0, str(Path(__file__).parent / "../../../bindings/python/client"))

try:
    from keystonedb import Client
    from keystonedb.builders import PutRequestBuilder, GetRequestBuilder, QueryRequestBuilder
except ImportError as e:
    print(f"Error importing keystonedb client: {e}")
    print("\nMake sure you've generated the protobuf files:")
    print("  python -m grpc_tools.protoc \\")
    print("    --python_out=bindings/python/client/keystonedb \\")
    print("    --grpc_python_out=bindings/python/client/keystonedb \\")
    print("    --proto_path=kstone-proto/proto \\")
    print("    kstone-proto/proto/keystone.proto")
    sys.exit(1)


SERVER_ADDR = "localhost:50051"


async def create_tasks(client: Client):
    """Create sample tasks in the database."""
    print("--- Creating Tasks ---")

    tasks = [
        {
            "id": "task#1",
            "project": "project#backend",
            "title": "Implement user authentication",
            "description": "Add JWT-based auth system",
            "status": "in-progress",
            "priority": "high",
        },
        {
            "id": "task#2",
            "project": "project#backend",
            "title": "Set up database migrations",
            "description": "Create migration scripts",
            "status": "pending",
            "priority": "medium",
        },
        {
            "id": "task#3",
            "project": "project#frontend",
            "title": "Design login page",
            "description": "Create UI mockups",
            "status": "completed",
            "priority": "low",
        },
    ]

    for task in tasks:
        # Build item attributes
        attributes = {
            "title": {"S": task["title"]},
            "description": {"S": task["description"]},
            "status": {"S": task["status"]},
            "priority": {"S": task["priority"]},
            "created": {"N": str(int(time.time()))},
        }

        # Put task by ID
        request = (
            PutRequestBuilder()
            .partition_key(task["id"].encode())
            .attributes(attributes)
            .build()
        )
        await client.put(request)

        # Also put with project partition for querying
        request_project = (
            PutRequestBuilder()
            .partition_key(task["project"].encode())
            .sort_key(task["id"].encode())
            .attributes(attributes)
            .build()
        )
        await client.put(request_project)

        print(f"✅ Created task: {task['id']}")


async def get_task(client: Client):
    """Retrieve a specific task."""
    print("\n--- Retrieving Task ---")

    request = GetRequestBuilder().partition_key(b"task#1").build()

    response = await client.get(request)

    if response.item:
        print("Task task#1:")
        print_item(response.item)
    else:
        print("Task not found")


async def query_tasks(client: Client):
    """Query tasks by project."""
    print("\n--- Querying Tasks by Project ---")

    request = (
        QueryRequestBuilder()
        .partition_key(b"project#backend")
        .limit(10)
        .build()
    )

    response = await client.query(request)

    print(f"Found {len(response.items)} tasks for project#backend")

    for i, item in enumerate(response.items, 1):
        print(f"\nTask {i}:")
        print_item(item)


async def batch_operations(client: Client):
    """Demonstrate batch get operation."""
    print("\n--- Batch Operations ---")

    from keystonedb.builders import BatchGetRequestBuilder

    request = (
        BatchGetRequestBuilder()
        .add_key(b"task#1")
        .add_key(b"task#2")
        .build()
    )

    response = await client.batch_get(request)

    print(f"Retrieved {len(response.items)} tasks in batch operation")


async def delete_task(client: Client):
    """Delete a task."""
    print("\n--- Deleting Task ---")

    from keystonedb.builders import DeleteRequestBuilder

    # Delete task#3
    request = DeleteRequestBuilder().partition_key(b"task#3").build()
    await client.delete(request)

    # Also delete from project partition
    request_project = (
        DeleteRequestBuilder()
        .partition_key(b"project#frontend")
        .sort_key(b"task#3")
        .build()
    )
    await client.delete(request_project)

    print("✅ Deleted task#3")


def print_item(item):
    """Pretty-print an item's attributes."""
    for key, value in item.attributes.items():
        # Extract value based on type
        val_str = None
        if value.HasField("S"):
            val_str = value.S
        elif value.HasField("N"):
            val_str = value.N
        elif value.HasField("Bool"):
            val_str = str(value.Bool)
        else:
            val_str = str(value)

        print(f"  {key}: {val_str}")


async def main():
    """Run all examples."""
    print(f"Connecting to KeystoneDB server at {SERVER_ADDR}...")

    async with Client(SERVER_ADDR) as client:
        print("✅ Connected successfully!\n")

        await create_tasks(client)
        await get_task(client)
        await query_tasks(client)
        await batch_operations(client)
        await delete_task(client)

        print("\n✅ All operations completed successfully!")


if __name__ == "__main__":
    import asyncio

    asyncio.run(main())
